hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndexWithSql.scala

```diff
@@ -34,74 +43,294 @@ class TestPartitionStatsIndexWithSql extends HoodieSparkSqlTestBase {
+        checkAnswer(s"select _hoodie_record_key, _hoodie_partition_path, id, name, price, ts, cast(dt as string) from $tableName order by id")(
+          Seq("1", s"dt=$partitionValue", 1, "a1", 11, 1001, partitionValue),
+          Seq("2", s"dt=$partitionValue", 2, "a2", 11, 1000, partitionValue)
+        )
+        // Test delete
+        spark.sql(s"delete from $tableName where id = 1")
+        checkAnswer(s"select _hoodie_record_key, _hoodie_partition_path, id, name, price, ts, cast(dt as string) from $tableName order by id")(
+          Seq("2", s"dt=$partitionValue", 2, "a2", 11, 1000, partitionValue)
+        )
+      }
+    }
+  }
+
+  test("Test partition stats index on string type field with insert and file pruning") {
+    Seq("cow", "mor").foreach { tableType =>
+      withTempDir { tmp =>
+        val tableName = generateTableName
+        val tablePath = s"${tmp.getCanonicalPath}/$tableName"
+        // Create table with date type partition
+        spark.sql(
+          s"""
+             | create table $tableName (
+             |    ts BIGINT,
+             |    uuid STRING,
+             |    rider STRING,
+             |    driver STRING,
+             |    city STRING,
+             |    state STRING
+             |) using hudi
+             | options(
+             |    type = '$tableType',
+             |    primaryKey ='uuid',
+             |    preCombineField = 'ts',
+             |    hoodie.metadata.index.partition.stats.enable = 'true',
+             |    hoodie.metadata.index.column.stats.column.list = 'rider'
+             |)
+             |PARTITIONED BY (state)
+             |location '$tablePath'
+         """.stripMargin
+        )
+        // set small file limit to 0 so that each insert creates a new file
+        spark.sql("set hoodie.parquet.small.file.limit=0")
+        // insert data in below pattern so that multiple records for 'texas' and 'california' partition are in same file
+        spark.sql(
+          s"""
+             | insert into $tableName
+             | values (1695159649,'334e26e9-8355-45cc-97c6-c31daf0df330','rider-A','driver-K','san_francisco','california'), (1695091554,'e96c4396-3fad-413a-a942-4cb36106d721','rider-F','driver-M','sunnyvale','california')
+             | """.stripMargin
+        )
+        spark.sql(s"INSERT INTO $tableName VALUES (1695332066,'7a84095f-737f-40bc-b62f-6b69664712d2','rider-B','driver-L','new york city','new york')")
+        spark.sql(s"INSERT INTO $tableName VALUES (1695516137,'3eeb61f7-c2b0-4636-99bd-5d7a5a1d2c04','rider-D','driver-M','princeton','new jersey')")
+        spark.sql(
+          s"""
+             | insert into $tableName
+             | values (1695516137,'e3cf430c-889d-4015-bc98-59bdce1e530c','rider-C','driver-P','houston','texas'), (1695332066,'1dced545-862b-4ceb-8b43-d2a568f6616b','rider-E','driver-O','austin','texas')
+             | """.stripMargin
+        )
```

Comment:
with `hoodie.parquet.small.file.limit` only product one file.

Reply:
@KnightChess Thanks for debugging the issue. Do you mind firing a fix for this issue?

Reply:
Regardless, I'm able to reproduce the problem based on your changes.